[Security Solution][Detections]Adds dry_run mode description to _bulk_action API (#2210)
elastic/security-docs: docs/detections/api/rules/rules-api-bulk-actions.asciidoc

`POST  <kibana host>:<port>/api/detection_engine/rules/_bulk_action`

[discrete]
===== URL query parameters

[width="100%",options="header"]
|==============================================
|Name |Type |Description |Required

|`dry_run` |Boolean | Enables dry run mode for the request call

> |No

Reviewer: @banderror this is how tables in Asciidoc are formatted -- each column is a on a new line. It looks OK in the preview.

> |No

Reviewer 2: @banderror here's the HTML preview for the new "Dry run mode" section: https://security-docs_2210.docs-preview.app.elstc.co/guide/en/security/master/bulk-actions-rules-api.html#_dry_run_mode

Preview of changes made to the "Bulk action" section: https://security-docs_2210.docs-preview.app.elstc.co/guide/en/security/master/bulk-actions-rules-api.html#bulk-actions-rules-api-action

Yes, if action is `edit`.
|==============================================

[discrete]
==== Dry run mode
In dry run mode (when URL query parameter `dry_run=true`) rules will be validated depends on action, but rules won't be saved in Elasticsearch.

> IMPORTANT: Because `export` action doesn't mutate state of rules and returns response in `.ndjson` file, dry run mode is not supported for this action

Author: simplified this bit

Yes, if action is `edit`.
|==============================================

[[bulk-actions-rules-api-dry-run]]
[discrete]
==== Dry run mode

> Enable dry run mode before you bulk update or delete rules to verify that the selected rules can be modified _before_ you permanently change them. When dry mode is enabled, the bulk action is temporarily applied to selected rules. These updates are not written to {es}.

Reviewer 2: @banderror @vitaliidm should we tweak the first sentence a bit more? Looking at this again, I'm not sure that it fully encompasses the main use cases that this feature addresses. Here's my understanding:
- The dry run allows users do a test run of the bulk action they select (enable, disable, delete, etc.) to help them verify that the rules they specified can actually be updated. 
- Users can complete a pelimiary and temporary bulk rule update with this feature. This would give them an extra buffer layer, or another layer of checks, before they permanently alter their rules.

Reviewer 2: If the above is correct, here's another revision that includes both. Let me know what you think:

```suggestion
Enable dry run mode to allow test runs when applying bulk actions to rules. Dry runs can help you verify that the rules you specified can actually be updated. They're also useful for testing bulk rule updates before they're permanently applied. When dry mode is enabled, bulk actions are temporarily applied to specified rules to show you a realistic set of results. Rule updates are not written to {es}.
```

Reviewer 3: @nastasha-solomon I don't think that strictly speaking this is a correct statement:

> When dry mode is enabled, bulk actions are temporarily applied to specified rules to show you a realistic set of results.

In the dry run mode, we don't return any results in the response body. We only return errors if there are any. So it's correct to say that this mode "can help you verify that the rules you specified can actually be updated". But it won't help the user to preview what would be the result (the updated, created, or deleted rules) of an action. So, you can think about this mode as a validation function.

Reviewer 2: @banderror ah, thanks for clarifying that. Would it be correct to say that dry run mode identifies rules that will not accept bulk actions?

Reviewer 3: @nastasha-solomon yep, that would be correct!

> Enable dry run mode to allow test runs when applying bulk actions to rules. Dry runs can help you verify that the rules you specified can actually be updated. They're also useful for testing bulk rule updates before they're permanently applied. When dry mode is enabled, bulk actions are applied to specified rules to show you a realistic set of results. Rule updates are not written to {es}.

Reviewer 2: ```suggestion
Enable dry run mode to verify that bulk actions can be applied to specified rules. Certain rules, such as prebuilt Elastic rules, can't be edited and will return errors in the request response. Error details will contain an explanation, the rule name and/or ID, and additional troubleshooting information. 

To enable dry run mode on a request, add the query parameter `dry_run=true` to the end of the request URL. Rules specified in the request will be temporarily updated. These updates won't be written to {es}.
```

Reviewer 2: @vitaliidm @banderror hopefully this revision is more accurate. Feel free to tweak it as needed!

Also, if you end up using this, you can remove the tip in line 284. I included it in the revision above.

Reviewer 3: That version looks perfect 🙏

> TIP: If `dry_run=true` is present at the end of the URL query parameter, rules specified in the request will be run in dry mode. 

Reviewer 2: ```suggestion
TIP: Add the query parameter `dry_run=true` to the end of the URL to enable dry run mode. Rules specified in the request will be run in dry run mode.
```

Reviewer 2: Here's one more revision if you want to make the tip actionable.

> IMPORTANT: Dry run mode is not supported for `export` action.

Reviewer 2: ```suggestion
IMPORTANT: Dry run mode is not supported for the `export` bulk action.
```

> IMPORTANT: Dry run mode is not supported for the `export` bulk action. Request's response will return `400` error.

Reviewer 2: ```suggestion
IMPORTANT: Dry run mode is not supported for the `export` bulk action. A `400` error will be returned in the request response.
```

Yes, if action is `edit`.
|==============================================

[[bulk-actions-rules-api-dry-run]]
[discrete]
==== Dry run mode
Enable dry run mode to verify that bulk actions can be applied to specified rules. Certain rules, such as prebuilt Elastic rules, can't be edited and will return errors in the request response. Error details will contain an explanation, the rule name and/or ID, and additional troubleshooting information. 

To enable dry run mode on a request, add the query parameter `dry_run=true` to the end of the request URL. Rules specified in the request will be temporarily updated. These updates won't be written to {es}.

> TIP: Add the query parameter `dry_run=true` to the URL to enable dry run mode. Rules specified in the request will be processed in dry run mode.

Reviewer 2: ```suggestion

```

Reviewer 2: This info is provided in line 284 so you can remove the tip :)

--------------------------------------------------

<1> dev:[] `execution_summary` is under development and its schema may change.

*Example 3, Dry run*

> The following request validates `add_index_patterns` action to the rules, which IDs included in the payload.
> Because, it's in dry ryn mode: no changes will be saved for these rules.

Reviewer 2: ```suggestion
The following request will validate that the `add_index_patterns` bulk action can be successfully applied to three rules. Each rule (which specified by its rule ID) is different: one is a prebuilt Elastic rule, another is a custom machine learning rule, and another is a custom query rule. Because dry run mode is enabled, changes to these rules will not be permanently applied and saved to {es}.
```

[source,console]
--------------------------------------------------
POST api/detection_engine/rules/_bulk_action?dry_run=true
{
    "action": "edit",
    "edit": [
        {
            "value": [
                "test-*"
            ],
            "type": "add_index_patterns"
        }
    ],
    "ids": ["81aa0480-06af-11ed-94fb-dd1a0597d8d2", "dc015d10-0831-11ed-ac8b-05a222bd8d4a", "de8f5af0-0831-11ed-ac8b-05a222bd8d4a"]
}
--------------------------------------------------

[discrete]
===== Response code

`500`::

>     Indicates partial bulk action failure.

Reviewer 3: ```suggestion
    Indicates a partial bulk action failure.
```

Author: I don't think  indefinite article is required here. cc @nastasha-solomon

Reviewer 2: Both versions of this work! That said, I have noticed that indefinite articles are commonly used throughout this topic (see [here](https://www.elastic.co/guide/en/security/master/bulk-actions-rules-api.html#_response_code_9) for an example) so I'd include it for consistency if nothing else.

[discrete]
===== Response payload

> Response payload returns 2 failed rules and 1 successful. In `attributes.errors` section of response, can be found error messages with list of failed rules(id and name).

Reviewer 3: ```suggestion
The response contains 2 failed and 1 successful result of running this test bulk action. In the `attributes.errors` section of the response, you can find the error messages and rules that failed the dry run validation. This means that if you would send the same request without `dry_run=true`, the same rules would fail.
```

Reviewer 2: ```suggestion
The `attributes.errors` section of the response shows that two rules failed to update and one succeeded. The same results would be returned if you ran the same request without dry run mode enabled.
```

> Contrary to the above example, that returns updated rule objects in `attributes.results`, in case of dry run mode, no rules objects will be returned, as rules are not getting saved in Elasticsearch.

Reviewer 3: ```suggestion
Notice that in the `attributes.results` all the arrays are empty. In the dry run mode, the endpoint doesn't return any `updated`, `created`, or `deleted` rules in the results, because no updates are being applied and saved to Elasticsearch.
```

Reviewer 2: ```suggestion
Notice that there are no arrays in `attributes.results`. In the dry run mode, rule updates are not applied and saved to {es}, which means the endpoint won't return results for rules that have been `updated`, `created`, or `deleted`.
```

--------------------------------------------------

<1> dev:[] `execution_summary` is under development and its schema may change.

*Example 3, Dry run*

> The following request will validate that the `add_index_patterns` bulk action can be successfully applied to three rules. Each rule (which specified by its rule ID) is different: one is a prebuilt Elastic rule, another is a custom machine learning rule, and another is a custom query rule. Because dry run mode is enabled, changes to these rules will not be permanently applied and saved to {es}.

Reviewer 4: ```suggestion
The following request will validate that the `add_index_patterns` bulk action can be successfully applied to three rules. Each rule (specified by its rule ID) is different: one is a prebuilt Elastic rule, another is a custom machine learning rule, and another is a custom query rule. Because dry run mode is enabled, changes to these rules will not be permanent or saved to {es}.
```

[source,console]
--------------------------------------------------
POST api/detection_engine/rules/_bulk_action?dry_run=true
{
    "action": "edit",
    "edit": [
        {
            "value": [
                "test-*"
            ],
            "type": "add_index_patterns"
        }
    ],
    "ids": ["81aa0480-06af-11ed-94fb-dd1a0597d8d2", "dc015d10-0831-11ed-ac8b-05a222bd8d4a", "de8f5af0-0831-11ed-ac8b-05a222bd8d4a"]
}
--------------------------------------------------

[discrete]
===== Response code

`500`::
    Indicates partial bulk action failure.

[discrete]
===== Response payload

> The `attributes.errors` section of the response shows that two rules failed to update and one succeeded. The same results would be returned if you ran the same request without dry run mode enabled.

Reviewer 2: ```suggestion
The `attributes.errors` section of the response shows that two rules failed to update and one succeeded. The same results would be returned if you ran the request without dry run mode enabled.
```

> Notice that there are no arrays in `attributes.results`. In the dry run mode, rule updates are not applied and saved to {es}, which means the endpoint won't return results for rules that have been `updated`, `created`, or `deleted`.

Reviewer 2: ```suggestion
Notice that there are no arrays in `attributes.results`. In the dry run mode, rule updates are not applied and saved to {es}, so the endpoint wouldn't return results for rules that have been `updated`, `created`, or `deleted`.
```

Reviewer 4: I actually like "won't" better than "wouldn't" here. `Are not... do not` seems more consistent to me than `are not... would not`

Reviewer 2: I agree! I'm also not a fan of using future tense verbs but couldn't think of an acceptable alternative. Thanks, @benironside !

> Notice that there are no arrays in `attributes.results`. In the dry run mode, rule updates are not applied and saved to {es}, which means the endpoint won't return results for rules that have been `updated`, `created`, or `deleted`.

Reviewer 4: ```suggestion
Notice that there are no arrays in `attributes.results`. In dry run mode, rule updates are not applied and saved to {es}, which means the endpoint won't return results for rules that have been `updated`, `created`, or `deleted`.
```
